File: core/src/mods/cleanup.rs
use std::fs;
use std::path::Path;
use std::process::Command;

use crate::mods::constants::{NGINX_SITES_AVAILABLE, NGINX_SITES_ENABLED};
use crate::mods::logger::{log_info, log_step, log_success, log_warning};

pub fn detect_broken_configs() -> Result<Vec<String>, String> {
    let mut broken = vec![];

    // Test nginx config
    let output = Command::new("nginx")
        .arg("-t")
        .output()
        .map_err(|e| format!("Failed to run nginx -t: {}", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        
        // Parser le output pour trouver les fichiers problématiques
        for line in stderr.lines() {
            // Chercher les patterns comme "cannot load certificate" ou "unknown directive"
            if line.contains("in /etc/nginx/sites-enabled/") {
                if let Some(start) = line.find("/etc/nginx/sites-enabled/") {
                    if let Some(domain_end) = line[start + 25..].find(':') {
                        let domain = &line[start + 25..start + 25 + domain_end];
                        if !broken.contains(&domain.to_string()) {
                            broken.push(domain.to_string());
                        }
                    }
                }
            } else if line.contains("cannot load certificate") {
                // Extraire le domaine du path du certificat
                if let Some(start) = line.find("/etc/letsencrypt/live/") {
                    if let Some(end) = line[start + 22..].find('/') {
                        let domain = &line[start + 22..start + 22 + end];
                        // Trouver la config correspondante
                        let config_path = format!("{}/{}", NGINX_SITES_ENABLED, domain);
                        if Path::new(&config_path).exists() && !broken.contains(&domain.to_string()) {
                            broken.push(domain.to_string());
                        }
                    }
                }
            }
        }
    }

    Ok(broken)
}

pub fn clean_broken_configs(dry_run: bool) -> Result<(), String> {
    log_step("🧹 Cleaning broken configurations...\n");

    let broken = detect_broken_configs()?;

    if broken.is_empty() {
        log_success("✓ No broken configurations found");
        return Ok(());
    }

    log_warning(&format!("Found {} broken configuration(s):", broken.len()));
    for domain in &broken {
        log_info(&format!("   - {}", domain));
    }

    if dry_run {
        log_warning("\nDry run mode: no changes made");
        return Ok(());
    }

    log_step("\n🗑️  Removing broken configurations...");
    for domain in &broken {
        match remove_config_files(domain) {
            Ok(_) => log_success(&format!("   ✓ Removed: {}", domain)),
            Err(e) => log_warning(&format!("   ⚠️  Failed to remove {}: {}", domain, e)),
        }
    }

    log_success("\n✅ Cleanup complete!");
    Ok(())
}

pub fn remove_config_files(domain: &str) -> Result<(), String> {
    let available_path = format!("{}/{}", NGINX_SITES_AVAILABLE, domain);
    let enabled_path = format!("{}/{}", NGINX_SITES_ENABLED, domain);

    if Path::new(&enabled_path).exists() {
        fs::remove_file(&enabled_path).map_err(|e| format!("Failed to remove symlink: {}", e))?;
    }

    if Path::new(&available_path).exists() {
        fs::remove_file(&available_path).map_err(|e| format!("Failed to remove config: {}", e))?;
    }

    Ok(())
}

File: core-go/engine/cleanup.go
package engine

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"xynginc/constants"
	"xynginc/logger"
)

func detectBrokenConfigs() ([]string, error) {
	var broken []string

	cmd := exec.Command("nginx", "-t")
	out, err := cmd.CombinedOutput()
	if err != nil {
		stderr := string(out)

		for _, line := range strings.Split(stderr, "\n") {
			if strings.Contains(line, "in /etc/nginx/sites-enabled/") {
				start := strings.Index(line, "/etc/nginx/sites-enabled/")
				if start != -1 {
					domainEnd := strings.Index(line[start+25:], ":")
					if domainEnd != -1 {
						domain := line[start+25 : start+25+domainEnd]
						if !contains(broken, domain) {
							broken = append(broken, domain)
						}
					}
				}
			} else if strings.Contains(line, "cannot load certificate") {
				start := strings.Index(line, "/etc/letsencrypt/live/")
				if start != -1 {
					end := strings.Index(line[start+22:], "/")
					if end != -1 {
						domain := line[start+22 : start+22+end]
						configPath := filepath.Join(constants.NginxSitesEnabled, domain)
						if _, err := os.Stat(configPath); !os.IsNotExist(err) && !contains(broken, domain) {
							broken = append(broken, domain)
						}
					}
				}
			}
		}
	}

	return broken, nil
}

func CleanBrokenConfigs(dryRun bool) error {
	logger.Step("🧹 Cleaning broken configurations...\n")

	broken, err := detectBrokenConfigs()
	if err != nil {
		return err
	}

	if len(broken) == 0 {
		logger.Success("✓ No broken configurations found")
		return nil
	}

	logger.Warning(fmt.Sprintf("Found %d broken configuration(s):", len(broken)))
	for _, domain := range broken {
		logger.Info(fmt.Sprintf("   - %s", domain))
	}

	if dryRun {
		logger.Warning("\nDry run mode: no changes made")
		return nil
	}

	logger.Step("\n>  Removing broken configurations...")
	for _, domain := range broken {
		if err := removeConfigFiles(domain); err != nil {
			logger.Warning(fmt.Sprintf("   ⚠️  Failed to remove %s: %v", domain, err))
		} else {
			logger.Success(fmt.Sprintf("   ✓ Removed: %s", domain))
		}
	}

	logger.Success("\n✅ Cleanup complete!")
	return nil
}

func removeConfigFiles(domain string) error {
	availablePath := filepath.Join(constants.NginxSitesAvailable, domain)
	enabledPath := filepath.Join(constants.NginxSitesEnabled, domain)

	if _, err := os.Stat(enabledPath); !os.IsNotExist(err) {
		if err := os.Remove(enabledPath); err != nil {
			return fmt.Errorf("failed to remove symlink: %v", err)
		}
	}

	if _, err := os.Stat(availablePath); !os.IsNotExist(err) {
		if err := os.Remove(availablePath); err != nil {
			return fmt.Errorf("failed to remove config: %v", err)
		}
	}

	return nil
}

func contains(slice []string, item string) bool {
	for _, a := range slice {
		if a == item {
			return true
		}
	}
	return false
}
